File: SpinEditor/CustomDisplayText/CustomDisplayTextVB/MyRadSpinElement.vb
Imports System.Globalization
Imports Telerik.WinControls.UI

#Region "CustomRadSpinElement"
Public Class MyRadSpinElement
    Inherits RadSpinElement

    Private _leadingZero As Boolean
    Private _scientificNation As Boolean

    Protected Overrides ReadOnly Property ThemeEffectiveType As Type
        Get
            Return GetType(RadSpinElement)
        End Get
    End Property

    Public Property ScientificNation As Boolean
        Get
            Return Me._scientificNation
        End Get
        Set(ByVal value As Boolean)

            If Me._scientificNation <> value Then
                Me._scientificNation = value
                Me.SetSpinValue(Me.internalValue, True)
            End If
        End Set
    End Property

    Public Property LeadingZero As Boolean
        Get
            Return Me._leadingZero
        End Get
        Set(ByVal value As Boolean)

            If Me._leadingZero <> value Then
                Me._leadingZero = value
                Me.SetSpinValue(Me.internalValue, True)
            End If
        End Set
    End Property

    Protected Overrides Function GetValueFromText() As Decimal
        If Not Me.ScientificNation Then
            Return MyBase.GetValueFromText()
        End If

        Try

            If Not String.IsNullOrEmpty(Me.Text) AndAlso ((Me.Text.Length <> 1) OrElse (Me.Text <> "-")) Then
                Return Me.Constrain(Decimal.Parse(Me.Text, NumberStyles.AllowExponent Or NumberStyles.AllowDecimalPoint))
            Else
                Return Me.internalValue
            End If

        Catch
            Return Me.internalValue
        End Try
    End Function

    Protected Overrides Sub SetSpinValue(ByVal value As Decimal, ByVal fromValue As Boolean)
        MyBase.SetSpinValue(value, fromValue)
        Me.TextBoxControl.Text = GetNumberText(Me.internalValue)
    End Sub

    Protected Overrides Function GetNumberText(ByVal num As Decimal) As String
        If Me.Hexadecimal Then
            Return String.Format("{0:X}", CLng(num))
        End If

        If Me.ScientificNation Then
            Return num.ToString("E", CultureInfo.CurrentCulture)
        End If

        If Me.LeadingZero Then
            Return num.ToString("00.##", CultureInfo.CurrentCulture)
        End If

        Return num.ToString((If(Me.ThousandsSeparator, "N", "F")) & Me.DecimalPlaces.ToString(CultureInfo.CurrentCulture), CultureInfo.CurrentCulture)
    End Function
End Class
#End Region

File: SpinEditor/CustomDisplayText/CustomDisplayText/MyRadSpinElement.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Telerik.WinControls.UI;

namespace CustomDisplayText
{
    #region CustomRadSpinElement
    public class MyRadSpinElement : RadSpinElement
    {
        private bool leadingZero;
        private bool scientificNation;

        protected override Type ThemeEffectiveType
        {
            get
            {
                return typeof(RadSpinElement);
            }
        }

        public bool ScientificNation
        {
            get
            {
                return this.scientificNation;
            }
            set
            {
                if (this.scientificNation != value)
                {
                    this.scientificNation = value;
                    this.SetSpinValue(this.internalValue, true);
                }
            }
        }

        public bool LeadingZero
        {
            get
            {
                return this.leadingZero;
            }
            set

            {
                if (this.leadingZero != value)
                {
                    this.leadingZero = value;
                    this.SetSpinValue(this.internalValue, true);
                }
            }
        }

        protected override decimal GetValueFromText()
        {
            if (!this.ScientificNation)
            {
                return base.GetValueFromText();
            }

            try
            {
                if (!string.IsNullOrEmpty(this.Text) && ((this.Text.Length != 1) || (this.Text != "-")))
                {

                    return this.Constrain(decimal.Parse(this.Text, NumberStyles.AllowExponent | NumberStyles.AllowDecimalPoint));
                }
                else
                {
                    return this.internalValue;
                }
            }
            catch
            {
                return this.internalValue;
            }
        }

        protected override void SetSpinValue(decimal value, bool fromValue)
        {
            base.SetSpinValue(value, fromValue);

            this.TextBoxControl.Text = GetNumberText(this.internalValue);
        }

        protected override string GetNumberText(decimal num)
        {
            if (this.Hexadecimal)
            {
                return string.Format("{0:X}", (long)num);
            }

            if (this.ScientificNation)
            {
                return num.ToString("E", CultureInfo.CurrentCulture);
            }

            if (this.LeadingZero)
            {
                return num.ToString("00.##", CultureInfo.CurrentCulture);
            }

            return num.ToString((this.ThousandsSeparator ? "N" : "F") + this.DecimalPlaces.ToString(CultureInfo.CurrentCulture), CultureInfo.CurrentCulture);
        }
    }

    #endregion
}
